File: 프로그래머스/2/42584. 주식가격/주식가격.py
def solution(prices):
    answer = []
    stk = []
    period = [len(prices)-i-1 for i in range(len(prices))]
    for i,c in enumerate(prices):
        if stk and stk[-1][1] > c:
            while stk and stk[-1][1] > c:
                ii,cc = stk.pop()
                period[ii] = i-ii
        stk.append((i,c))
    return period

File: 프로그래머스/2/42584. 주식가격/주식가격.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

vector<int> solution(vector<int> prices) {
    vector<int> answer(prices.size());
    vector<int> stk;
    
    for(int i = 0; i < prices.size(); i++){
        while(!stk.empty() && prices[stk.back()] > prices[i]){
            int p = stk.back();
            stk.pop_back();
            answer[p] = i - p;
        }
        stk.push_back(i);
    }
    while(!stk.empty()){
        int p = stk.back();
        stk.pop_back();
        answer[p] = (prices.size()-1)-p;
    }
    return answer;
}

// 1 2 3 